### PR TITLE
Add "Alpine Coaster" ride type

### PR DESCRIPTION
**Pull request by X123M3-256**
This PR adds a new ride type "Alpine Coaster" which is a Wiegand alpine coaster. It includes gentle slopes, banked turns and sloped turns, but not banked sloped turns. I omitted the latter because I would like to implement some more sloped track pieces for this ride in future, and having banked sloped turns available would massively inflate the number of sprites required to do that.

alpineprimage

In addition, a new vehicle flag "RIDE_ENTRY_FLAG_RIDER_CONTROLS_SPEED" has been implemented. This is intended to be set on rides where the riders can control the speed of their vehicle (of which this is one). There are 64 presets for the rider behaviour, of which one is randomly chosen each time a car dispatches. I have overloaded one of the vehicle fields to store this value; it shouldn't cause any problems as I believe that field is only used for swinging rides. 

The rider behaviour is determined by four values - their preferred speed on straight track (above which they will apply braking to slow down), their preferred speed in turns , their tolerance for going faster than their preferred speed (i.e how much faster they can go before they slam on the brakes), and the distance they will leave between them and the rider in front. The peep speeds follow a normal distribution, so that particularly fast or slow peeps are less common than those near the middle of the range, so there aren't too many slow peeps holding everyone up.

I have also implemented a new ride flag "RIDE_TYPE_FLAG_UP_SLOPE_REQUIRES_LIFT", which disables the building of uphill track without a lift. In the vanilla game, rides like the log flume and splash boats simply replace the uphill slopes with lift sprites, and make the vehicle powered. I considered following that approach here, however, because this ride has curved slopes available, the player would then be able to build a curved lift, which I don't want. I also don't want to allow uphill slopes to be built without cheats, because the randomized nature of the rider braking means that a ride that appears to work during testing will almost certainly get stuck sooner or later, if it has any uphill slopes. Therefore, I felt it was necessary to add this new flag.

### CODE REVIEW

File: src/openrct2/ride/coaster/AlpineCoaster.cpp
Comment (oli414) [inline src/openrct2/ride/coaster/AlpineCoaster.cpp:1]: 2022

**Comment by duncanspumpkin** [comment]
Remove all of the `RIDE_TYPE_FLAG_UP_SLOPE_REQUIRES_LIFT` changes they should be done in a separate PR. This should be purely the new coaster i.e. track paint, meta, images.

**Comment by duncanspumpkin** [comment]
Remove `RIDE_TYPE_FLAG_UP_SLOPE_REQUIRES_LIFT` changes and make them as a separate PR.

**Comment by duncanspumpkin** [comment]
You seem to have forgotten to commit the meta file for the ride as well.

**Comment by X123M3-256** [comment]
@duncanspumpkin I have submitted #16835 and #16837, I think it will be easiest to wait for those to be merged and then rebase this branch.

**Comment by Dubaifreak** [comment]
This ride would be cool, i can use it when it's accepted!

**Comment by X123M3-256** [comment]
> Unless I am mistaken, there is no ./meta/AlpineCoaster.h file yet, is there?

Yes I forgot to commit the meta file, but this branch will need to be rebased once my other two PRs are merged so I'll sort it then.

**Comment by X123M3-256** [comment]
@duncanspumpkin I've rebased this branch and added the missing meta file. I'll submit a PR for the vehicle object shortly.

**Comment by duncanspumpkin** [comment]
@X123M3-256  i'm going to merge #16940 if you could rebase once more ontop of dev after that and adjust your prices that would be good.

**Comment by duncanspumpkin** [comment]
Could you provide info on which ride type you based your drawing code on. Did you require anything custom in the drawing code? Just for future reference when we try join together the different drawing code to make custom ride types work.

**Comment by X123M3-256** [comment]
> Could you provide info on which ride type you based your drawing code on. Did you require anything custom in the drawing code? Just for future reference when we try join together the different drawing code to make custom ride types work.

I copied the paint code from the single rail coaster and just changed the support style. IIRC, the single rail track paint code was originally based on the giga coaster, but there were changes made to the bounding boxes.

I think as far as custom ride types go, it will be necessary to be able to specify different bounding boxes for different ride types, as I find these often have to be adjusted for each specific ride type to get them to draw without glitching. However, the rest of the draw code is mostly identical between ride types. I think the code could be refactored so that you have one set of paint functions for all basic steel coasters, another for inverted coasters, and another for rides like the bobsled and side friction where the track is drawn in two halves.

**Comment by duncanspumpkin** [comment]
> > Could you provide info on which ride type you based your drawing code on. Did you require anything custom in the drawing code? Just for future reference when we try join together the different drawing code to make custom ride types work.
> 
> I copied the paint code from the single rail coaster and just changed the support style. IIRC, the single rail track paint code was originally based on the giga coaster, but there were changes made to the bounding boxes.
> 
> I think as far as custom ride types go, it will be necessary to be able to specify different bounding boxes for different ride types, as I find these often have to be adjusted for each specific ride type to get them to draw without glitching. However, the rest of the draw code is mostly identical between ride types. I think the code could be refactored so that you have one set of paint functions for all basic steel coasters, another for inverted coasters, and another for rides like the bobsled and side friction where the track is drawn in two halves.

The way it works in Locomotion is you have a number of different paint styles for an object. I was imaging we would have similar probably around 4 styles from my investigation for coasters. A style has to have matching number of sprites (although empty sprites are fine). In Locomotion you can't select your bound boxes for the style but yes I think we will allow that for OpenRCT2 due to the more intricate nature. We would then only have the 4 versions of drawing every track piece (okay maybe not all as some styles it might not make sense).

**Comment by X123M3-256** [comment]
@duncanspumpkin 

You might be able to get away without having custom bounding boxes for every track style, and instead have a set of different ones for different sizes of track that you can select between, a bit like how the vehicle objects work.

File: src/openrct2/ride/coaster/meta/AlpineCoaster.h
Comment (karst) [inline src/openrct2/ride/coaster/meta/AlpineCoaster.h:2]:
```suggestion
 * Copyright (c) 2014-2022 OpenRCT2 developers
```

File: src/openrct2/ride/coaster/AlpineCoaster.cpp
Comment (karst) [inline src/openrct2/ride/coaster/AlpineCoaster.cpp]:
```suggestion
#pragma region Copyright(c) 2014 - 2022 OpenRCT2 Developers
```

I noticed you marked Oli's conversation as resolved, but didn't update the copyright to 2022, I don't know if this was intentional or not.

Comment (X123M3-256) [reply src/openrct2/ride/coaster/AlpineCoaster.cpp]: I did update it?

Comment (karst) [reply src/openrct2/ride/coaster/AlpineCoaster.cpp]: Ah, it has now!

Comment (Broxzier) [reply src/openrct2/ride/coaster/AlpineCoaster.cpp]: Indeed, the full diff shows 2022 instead of 2017. You may have looked at a specific commit, @karst.

Comment (karst) [reply src/openrct2/ride/coaster/AlpineCoaster.cpp]: Yea I was just a bit too slow haha, got fixed a minute earlier in fcd9b73

**Comment by X123M3-256** [comment]
I have submitted a PR for the vehicle object to the object repo https://github.com/OpenRCT2/objects/pull/173.

**Comment by karst** [comment]
I noticed a changelog entry has not been made for this yet. Don't know when this will be merged though, but putting the reminder out there that it hasn't been done yet.

**Comment by karst** [comment]
I suddenly realized what the German and Dutch names are for alpine coasters. 
It's Rodelbahn in German and Rodelbaan in Dutch. So for the translation team that'll probably be something to add

**Review by duncanspumpkin — APPROVED** [review]
Code wise I'm happy with this but i leave graphics for others to approve.

**Comment by X123M3-256** [comment]
@duncanspumpkin I've rebased the branch but I can't find where the changelog is.

**Comment by X123M3-256** [comment]
@duncanspumpkin I've rebased the branch but I can't find where the changelog is.

**Comment by zacharyhub** [comment]
I wish that somebody make Schiff Wild Mouse track.

**Pull request merged**
